File: backend/soo/docx_reader.py
"""Read a .docx into a flat block stream, preserving list identity.

Uses only the stdlib. Word's numbering (`numId`) is what binds list items to the
lead-in sentence above them, so it has to survive into the block stream — that
association is the backbone of the whole parse.
"""
from __future__ import annotations

import re
import zipfile
from dataclasses import dataclass, field
from xml.etree import ElementTree as ET

W = "{http://schemas.openxmlformats.org/wordprocessingml/2006/main}"

# Not every author uses Word's list numbering. `016. TES Tanks` types the bullet
# glyph directly into an ordinary paragraph, which leaves no numId to group on, so
# those get a synthetic one. Contiguity does the rest of the work.
BULLET_RE = re.compile(r"^\s*[•▪●○◦‣⁃]\s+")
SYNTHETIC_NUM_ID = "bullet"


@dataclass
class Block:
    index: int
    kind: str  # "para" | "table"
    text: str = ""
    style: str = ""
    num_id: str | None = None
    level: int = 0
    rows: list[list[str]] = field(default_factory=list)

    @property
    def is_list_item(self) -> bool:
        return self.num_id is not None


def _text(el) -> str:
    return "".join(t.text or "" for t in el.iter(W + "t"))


def _style(p) -> str:
    pr = p.find(W + "pPr")
    if pr is None:
        return ""
    s = pr.find(W + "pStyle")
    return s.get(W + "val", "") if s is not None else ""


def _numbering(p) -> tuple[str | None, int]:
    pr = p.find(W + "pPr")
    if pr is None:
        return None, 0
    num = pr.find(W + "numPr")
    if num is None:
        return None, 0
    num_id = num.find(W + "numId")
    ilvl = num.find(W + "ilvl")
    return (
        num_id.get(W + "val") if num_id is not None else None,
        int(ilvl.get(W + "val")) if ilvl is not None else 0,
    )


def _table_rows(tbl) -> list[list[str]]:
    rows = []
    for tr in tbl.findall(W + "tr"):
        rows.append([_text(tc).strip() for tc in tr.findall(W + "tc")])
    return rows


def read_blocks(path: str) -> list[Block]:
    """Flatten a .docx body into ordered Blocks. Empty paragraphs are dropped."""
    with zipfile.ZipFile(path) as z:
        root = ET.fromstring(z.read("word/document.xml"))

    body = root.find(W + "body")
    if body is None:
        return []

    blocks: list[Block] = []
    for el in body:
        tag = el.tag.replace(W, "")
        if tag == "p":
            text = " ".join(_text(el).split())
            if not text:
                continue
            num_id, level = _numbering(el)
            if num_id is None and BULLET_RE.match(text):
                text = BULLET_RE.sub("", text)
                num_id = SYNTHETIC_NUM_ID
            blocks.append(
                Block(
                    index=len(blocks),
                    kind="para",
                    text=text,
                    style=_style(el),
                    num_id=num_id,
                    level=level,
                )
            )
        elif tag == "tbl":
            rows = _table_rows(el)
            if rows:
                blocks.append(Block(index=len(blocks), kind="table", rows=rows))

    return blocks

File: app/server/parse/docx-reader.ts
/**
 * .docx -> flat block stream, preserving list identity.
 *
 * Word's numbering (`numId`) is what binds list items to the lead-in sentence
 * above them, and that association is the backbone of the whole parse — so it has
 * to survive into the block stream. Document order matters too, hence
 * `preserveOrder` on the XML parser.
 */
import JSZip from "jszip";
import { XMLParser } from "fast-xml-parser";

export interface Block {
  index: number;
  kind: "para" | "table";
  text: string;
  style: string;
  numId: string | null;
  level: number;
  rows: string[][];
  isListItem: boolean;
}

/** Not every author uses Word's list numbering — some type the glyph directly. */
const BULLET_RE = /^\s*[•▪●○◦‣⁃]\s+/;
const SYNTHETIC_NUM_ID = "bullet";

type El = Record<string, any>;

const parser = new XMLParser({
  preserveOrder: true,
  ignoreAttributes: false,
  attributeNamePrefix: "@_",
  trimValues: false,
  // Word splits "PART 12" across runs, so the "12" arrives as its own text node.
  // Left on, the parser coerces it to a number and the digits vanish from the heading.
  parseTagValue: false,
  parseAttributeValue: false,
});

function tagOf(el: El): string {
  for (const k of Object.keys(el)) if (k !== ":@") return k;
  return "";
}

function kidsOf(el: El): El[] {
  const t = tagOf(el);
  const v = el[t];
  return Array.isArray(v) ? v : [];
}

function attrsOf(el: El): Record<string, string> {
  return (el[":@"] as Record<string, string>) ?? {};
}

function findChild(el: El, tag: string): El | undefined {
  return kidsOf(el).find((k) => tagOf(k) === tag);
}

/** Concatenate every `w:t` beneath this element, in order. */
function allText(el: El): string {
  const tag = tagOf(el);
  if (tag === "w:t") {
    return kidsOf(el)
      .map((k) => (k["#text"] === undefined || k["#text"] === null ? "" : String(k["#text"])))
      .join("");
  }
  if (tag === "w:tab") return " ";
  if (tag === "w:br") return " ";
  return kidsOf(el).map(allText).join("");
}

function styleOf(p: El): string {
  const pPr = findChild(p, "w:pPr");
  if (!pPr) return "";
  const pStyle = findChild(pPr, "w:pStyle");
  return pStyle ? (attrsOf(pStyle)["@_w:val"] ?? "") : "";
}

function numberingOf(p: El): { numId: string | null; level: number } {
  const pPr = findChild(p, "w:pPr");
  if (!pPr) return { numId: null, level: 0 };
  const numPr = findChild(pPr, "w:numPr");
  if (!numPr) return { numId: null, level: 0 };
  const numId = findChild(numPr, "w:numId");
  const ilvl = findChild(numPr, "w:ilvl");
  return {
    numId: numId ? (attrsOf(numId)["@_w:val"] ?? null) : null,
    level: ilvl ? Number(attrsOf(ilvl)["@_w:val"] ?? 0) : 0,
  };
}

function tableRows(tbl: El): string[][] {
  const rows: string[][] = [];
  for (const tr of kidsOf(tbl)) {
    if (tagOf(tr) !== "w:tr") continue;
    const cells: string[] = [];
    for (const tc of kidsOf(tr)) {
      if (tagOf(tc) !== "w:tc") continue;
      cells.push(allText(tc).replace(/\s+/g, " ").trim());
    }
    if (cells.length) rows.push(cells);
  }
  return rows;
}

function makeBlock(partial: Partial<Block> & { index: number; kind: Block["kind"] }): Block {
  return {
    text: "",
    style: "",
    numId: null,
    level: 0,
    rows: [],
    isListItem: false,
    ...partial,
  };
}

export async function readDocxBlocks(buffer: Buffer): Promise<Block[]> {
  const zip = await JSZip.loadAsync(buffer);
  const entry = zip.file("word/document.xml");
  if (!entry) throw new Error("Not a Word document: word/document.xml is missing");

  const xml = await entry.async("string");
  const tree = parser.parse(xml) as El[];

  const doc = tree.find((el) => tagOf(el) === "w:document");
  const body = doc ? findChild(doc, "w:body") : undefined;
  if (!body) return [];

  const blocks: Block[] = [];
  for (const el of kidsOf(body)) {
    const tag = tagOf(el);

    if (tag === "w:p") {
      let text = allText(el).replace(/\s+/g, " ").trim();
      if (!text) continue;
      let { numId, level } = numberingOf(el);
      if (numId === null && BULLET_RE.test(text)) {
        text = text.replace(BULLET_RE, "");
        numId = SYNTHETIC_NUM_ID;
      }
      blocks.push(
        makeBlock({
          index: blocks.length,
          kind: "para",
          text,
          style: styleOf(el),
          numId,
          level,
          isListItem: numId !== null,
        }),
      );
      continue;
    }

    if (tag === "w:tbl") {
      const rows = tableRows(el);
      if (rows.length) {
        blocks.push(makeBlock({ index: blocks.length, kind: "table", rows }));
      }
    }
  }

  return blocks;
}
